### PR TITLE
MMF-493 Make public settings Web API suitable for launching an analysis

### PR DESCRIPTION
**Pull request by julienlancelot**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/ValuesAction.java [hunk 1]
```diff
@@ -126,32 +129,36 @@ private static ValuesRequest toWsRequest(Request request) {
...
+      .collect(Collectors.toList());
+  }
+
+  private Predicate<Setting> isVisible(Optional<ComponentDto> component) {
+    return setting -> !setting.getKey().endsWith(SECURED_SUFFIX_KEY)
+      || hasAdminPermission(component)
+      || (setting.getKey().contains(LICENSE) && userSession.isLoggedIn());
+  }
+
+  private boolean hasAdminPermission(Optional<ComponentDto> component) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/ValuesAction.java]: MINOR Specify a "ComponentDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/ValuesAction.java [hunk 2]
```diff
@@ -104,54 +120,80 @@ public void handle(Request request, Response response) throws Exception {
...
+      .collect(Collectors.toList());
+  }
+
+  private Predicate<Setting> isVisible(Optional<ComponentDto> component) {
+    return setting -> !setting.getKey().endsWith(SECURED_SUFFIX)
+      || hasAdminPermission(component)
+      || (isLicenseRelated(setting) && userSession.isLoggedIn());
+  }
+
+  private boolean hasAdminPermission(Optional<ComponentDto> component) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/ValuesAction.java]: MINOR Specify a "ComponentDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/ValuesAction.java [hunk 3]
```diff
@@ -104,54 +115,78 @@ public void handle(Request request, Response response) throws Exception {
...
-    if (component.isPresent()) {
-      userSession.checkComponentUuidPermission(UserRole.ADMIN, component.get().uuid());
-    } else {
-      userSession.checkPermission(GlobalPermissions.SYSTEM_ADMIN);
+  private Stream<String> loadLicenseHashKeys() {
+    return propertyDefinitions.getAll()
+      .stream()
+      .filter(setting -> setting.type().equals(LICENSE))
+      .map(PropertyDefinition::key)
+      .map(setting -> setting.replace(LICENSE_SUFFIX, LICENSE_HASH_SUFFIX));
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/ValuesAction.java]: question: why the change in setting suffix key?

Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/ValuesAction.java]: suggestion: rename `setting` to `key`

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/ValuesAction.java [hunk 4]
```diff
@@ -104,54 +115,78 @@ public void handle(Request request, Response response) throws Exception {
...
+  }
+
+  private Optional<ComponentDto> loadComponent(DbSession dbSession, ValuesRequest valuesRequest) {
+    String componentKey = valuesRequest.getComponent();
+    if (componentKey != null) {
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/ValuesAction.java]: suggestion: either inline or start with
```
if(componentKey==null) {
  return Optional.empty();
}
```

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/ValuesAction.java [hunk 5]
```diff
@@ -104,54 +115,78 @@ public void handle(Request request, Response response) throws Exception {
...
-    }
-    return settings;
+    component.ifPresent(componentDto -> settings.addAll(settingsFinder.loadComponentSettings(dbSession, keys, componentDto).values()));
+    return settings.stream()
+      .filter(isVisible(component))
+      .collect(Collectors.toList());
+  }
+
+  private Predicate<Setting> isVisible(Optional<ComponentDto> component) {
+    return setting -> !setting.getKey().endsWith(SECURED_SUFFIX)
```
Comment (teryk) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/ValuesAction.java]: remark: don't you have the feeling our setting objects should be smarter? We should be able to have setting.isSecured() and encapsulate the logic
I know we use Protobuf objects, but we should try to find a better design if it's not directly with setting.isSecured()

**Comment by teryk** [comment]
technical review done

File: it/it-tests/src/test/java/it/webhook/WebhooksTest.java [hunk 6]
```diff
@@ -269,10 +269,10 @@ private void disableGlobalWebhooks() {
 
   private void setProperty(@Nullable String componentKey, String key, @Nullable String value) {
     if (value == null) {
-      ResetRequest req = ResetRequest.builder().setKeys(key).setComponentKey(componentKey).build();
```
Comment (julienlancelot) [inline it/it-tests/src/test/java/it/webhook/WebhooksTest.java]: Yep I can !

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/ListDefinitionsAction.java [hunk 7]
```diff
@@ -113,12 +108,11 @@ private static ListDefinitionsRequest toWsRequest(Request request) {
   }
 
   private Optional<ComponentDto> getComponent(DbSession dbSession, ListDefinitionsRequest wsRequest) {
-    String componentId = wsRequest.getComponentId();
-    String componentKey = wsRequest.getComponentKey();
-    if (componentId != null || componentKey != null) {
-      return Optional.of(componentFinder.getByUuidOrKey(dbSession, componentId, componentKey, ID_AND_KEY));
+    String componentKey = wsRequest.getComponent();
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/ListDefinitionsAction.java]: Good point, but we'll have a lot of issues as the Java plugin doesn't like public class to return Optional...

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/ValuesAction.java [hunk 8]
```diff
@@ -104,54 +115,78 @@ public void handle(Request request, Response response) throws Exception {
...
 
   private static ValuesRequest toWsRequest(Request request) {
-    return ValuesRequest.builder()
-      .setKeys(request.mandatoryParamAsStrings(PARAM_KEYS))
-      .setComponentId(request.param(PARAM_COMPONENT_ID))
-      .setComponentKey(request.param(PARAM_COMPONENT_KEY))
-      .build();
+    ValuesRequest.Builder builder = ValuesRequest.builder()
+      .setComponent(request.param(PARAM_COMPONENT));
+    if (request.hasParam(PARAM_KEYS)) {
```
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/ValuesAction.java]: There's no default value

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/ValuesAction.java (hunk 3 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/ValuesAction.java]: I can explain you offline

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/ValuesAction.java (hunk 5 above)
Comment (julienlancelot) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/ValuesAction.java]: I agree, we should improve the API to declare secured settings.

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingsPermissionPredicates.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+  Predicate<Setting> isSettingVisible(Optional<ComponentDto> component) {
+    return setting -> isVisible(setting.getKey(), component);
+  }
+
+  Predicate<PropertyDefinition> isDefinitionVisible(Optional<ComponentDto> component) {
+    return propertyDefinition -> isVisible(propertyDefinition.key(), component);
+  }
+
+  boolean isVisible(String settingKey, Optional<ComponentDto> component) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingsPermissionPredicates.java]: MINOR Specify a "ComponentDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingsPermissionPredicates.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    return !settingKey.endsWith(SECURED_SUFFIX)
+      || hasAdminPermission(component)
+      || (isLicenseRelated(settingKey) && userSession.isLoggedIn());
+  }
+
+  private boolean hasAdminPermission(Optional<ComponentDto> component) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingsPermissionPredicates.java]: MINOR Specify a "ComponentDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingsPermissionPredicates.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+  static final String LICENSE_SUFFIX = ".license.secured";
+  static final String LICENSE_HASH_SUFFIX = ".licenseHash.secured";
+
+  private final UserSession userSession;
+
+  public SettingsPermissionPredicates(UserSession userSession) {
+    this.userSession = userSession;
+  }
+
+  Predicate<Setting> isSettingVisible(Optional<ComponentDto> component) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingsPermissionPredicates.java]: MINOR Specify a "ComponentDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingsPermissionPredicates.java [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    return setting -> isVisible(setting.getKey(), component);
+  }
+
+  Predicate<PropertyDefinition> isDefinitionVisible(Optional<ComponentDto> component) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingsPermissionPredicates.java]: MINOR Specify a "ComponentDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/ListDefinitionsAction.java [hunk 13]
```diff
@@ -85,47 +87,35 @@ public void handle(Request request, Response response) throws Exception {
...
-  }
-
-  private Optional<ComponentDto> getComponent(DbSession dbSession, ListDefinitionsRequest wsRequest) {
-    String componentId = wsRequest.getComponentId();
-    String componentKey = wsRequest.getComponentKey();
-    if (componentId != null || componentKey != null) {
-      return Optional.of(componentFinder.getByUuidOrKey(dbSession, componentId, componentKey, ID_AND_KEY));
-    }
-    return Optional.empty();
+  private Optional<String> getQualifier(Optional<ComponentDto> component) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/ListDefinitionsAction.java]: MAJOR Make "getQualifier" a "static" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* MAJOR 1 major
* MINOR 4 minor

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingsPermissionPredicates.java [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+
+  public SettingsPermissionPredicates(UserSession userSession) {
+    this.userSession = userSession;
+  }
+
+  Predicate<Setting> isSettingVisible(Optional<ComponentDto> component) {
+    return setting -> isVisible(setting.getKey(), setting.getDefinition(), component);
+  }
+
+  Predicate<PropertyDefinition> isDefinitionVisible(Optional<ComponentDto> component) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingsPermissionPredicates.java]: MINOR Specify a "ComponentDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingsPermissionPredicates.java [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+    return propertyDefinition -> isVisible(propertyDefinition.key(), propertyDefinition, component);
+  }
+
+  private boolean isVisible(String key, @Nullable PropertyDefinition definition, Optional<ComponentDto> component) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingsPermissionPredicates.java]: MINOR Specify a "ComponentDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingsPermissionPredicates.java [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+    return verifySecuredSetting(key, definition, component) && (verifyLicenseSetting(key, definition));
+  }
+
+  private boolean verifySecuredSetting(String key, @Nullable PropertyDefinition definition, Optional<ComponentDto> component) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingsPermissionPredicates.java]: MINOR Specify a "ComponentDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingsPermissionPredicates.java [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+    return isLicense(key, definition) || (!key.endsWith(SECURED_SUFFIX) || hasAdminPermission(component));
+  }
+
+  private boolean hasAdminPermission(Optional<ComponentDto> component) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingsPermissionPredicates.java]: MINOR Specify a "ComponentDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/ListDefinitionsAction.java [hunk 18]
```diff
@@ -85,47 +87,35 @@ public void handle(Request request, Response response) throws Exception {
...
-  }
-
-  private Optional<ComponentDto> getComponent(DbSession dbSession, ListDefinitionsRequest wsRequest) {
-    String componentId = wsRequest.getComponentId();
-    String componentKey = wsRequest.getComponentKey();
-    if (componentId != null || componentKey != null) {
-      return Optional.of(componentFinder.getByUuidOrKey(dbSession, componentId, componentKey, ID_AND_KEY));
-    }
-    return Optional.empty();
+  private static Optional<String> getQualifier(Optional<ComponentDto> component) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/ListDefinitionsAction.java]: MAJOR Make "getQualifier" a "static" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues
* MAJOR 1 major
* MINOR 5 minor

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* MINOR 5 minor

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingsPermissionPredicates.java [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+
+  Predicate<Setting> isSettingVisible(Optional<ComponentDto> component) {
+    return setting -> isVisible(setting.getKey(), setting.getDefinition(), component);
+  }
+
+  Predicate<PropertyDefinition> isDefinitionVisible(Optional<ComponentDto> component) {
+    return propertyDefinition -> isVisible(propertyDefinition.key(), propertyDefinition, component);
+  }
+
+  boolean isVisible(String key, @Nullable PropertyDefinition definition, Optional<ComponentDto> component) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingsPermissionPredicates.java]: MINOR Specify a "ComponentDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)

File: server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingsPermissionPredicates.java [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+    return verifySecuredSetting(key, definition, component) && (verifyLicenseSetting(key, definition));
+  }
+
+  private boolean verifySecuredSetting(String key, @Nullable PropertyDefinition definition, Optional<ComponentDto> component) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/setting/ws/SettingsPermissionPredicates.java]: MINOR Specify a "ComponentDto" parameter instead. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3553)
